### PR TITLE
KAFKA-3490; Multiple version support for ducktape performance tests

### PR DESCRIPTION
**Pull request by ijuma**

### CODE REVIEW

**Comment by ijuma** [comment]
I revised @granders branch and rebased it against trunk. Running the branch builder here:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/401/

@granders, if you could take a look that would be great.

File: tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java [hunk 1]
```diff
@@ -234,6 +234,13 @@ public void printTotal() {
                               percs[1],
                               percs[2],
                               percs[3]);
+
+            /*      0: config.dateFormat.format(startMs), 1: config.dateFormat.format(endMs),
+      2: config.compressionCodec.codec, 3: config.messageSize, 4: config.batchSize, 5: totalMBSent,
+      6: mb/s, 7: totalMessagesSent.get, 8: msgs/sec))
+             * *
+             */
+            // 2015-11-17 03:46:42:901, 2015-11-17 03:46:43:158, 0, 100, 200, 0.00, 0.0037, 10, 38.9105\n',)}
```
Comment (ijuma) [inline tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java]: I will remove this.

File: tests/kafkatest/services/performance/end_to_end_latency.py [hunk 2]
```diff
@@ -36,17 +38,23 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, consumer_fetch
...
-            node.account.create_file(security_config_file, str(self.security_config))
         else:
             security_config_file = ""
+        return security_config_file
+
+    def start_cmd(self, node):
+        args = self.args.copy()
         args.update({
             'zk_connect': self.kafka.zk.connect_setting(),
             'bootstrap_servers': self.kafka.bootstrap_servers(self.security_config.security_protocol),
```
Comment (granders) [inline tests/kafkatest/services/performance/end_to_end_latency.py:65]: Line 61:
I think we want `'security_config_file': self.security_config_file`? `security_config_file` looks like it's out of scope now

**Comment by ijuma** [comment]
Started a system tests build here: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/405/

File: tests/kafkatest/services/performance/consumer_performance.py [hunk 3]
```diff
@@ -127,15 +135,31 @@ def start_cmd(self, node):
         cmd += " /opt/%s/bin/kafka-consumer-perf-test.sh" % kafka_dir(node)
         for key, value in self.args.items():
             cmd += " --%s %s" % (key, value)
-        cmd += " --consumer.config %s" % ConsumerPerformanceService.CONFIG_FILE
 
-        for key, value in self.settings.items():
-            cmd += " %s=%s" % (str(key), str(value))
```
Comment (ijuma) [inline tests/kafkatest/services/performance/consumer_performance.py]: @granders, is it right that these two lines should be removed?

File: tests/kafkatest/services/performance/producer_performance.py [hunk 4]
```diff
@@ -13,21 +13,45 @@
...
+            "collect_default": True},
+        "producer_performance_stderr": {
+            "path": STDERR_CAPTURE,
+            "collect_default": True},
         "producer_performance_log": {
-            "path": "/mnt/producer-performance.log",
+            "path": LOG_FILE,
             "collect_default": True},
+        "jmx_log": {
+            "path": "/mnt/jmx_tool.log",
```
Comment (ijuma) [inline tests/kafkatest/services/performance/producer_performance.py]: @granders, is it intentional that the jmx log is not under `PERSISTENT_ROOT`?

File: tests/kafkatest/services/performance/end_to_end_latency.py (hunk 2 above)
Comment (ijuma) [inline tests/kafkatest/services/performance/end_to_end_latency.py:65]: Fixed.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 4 above)
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: @ijuma I don't think so - maybe an artifact of cherry-picking?

Comment (ijuma) [inline tests/kafkatest/services/performance/producer_performance.py]: I think this is because `/mnt/jmx_tool.log` is hardcoded in `jmx.py`.

File: tests/kafkatest/services/performance/consumer_performance.py (hunk 3 above)
Comment (granders) [inline tests/kafkatest/services/performance/consumer_performance.py]: @ijuma Good catch - I agree these should not be removed

File: tests/kafkatest/sanity_checks/test_performance_services.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        self.record_size = 100
+        self.num_records = 10000
+        self.topic = "topic"
+
+        self.zk = ZookeeperService(test_context, 1)
+
+    def setUp(self):
+        self.zk.start()
+
+    @parametrize(version=str(LATEST_0_8_2))
```
Comment (granders) [inline tests/kafkatest/sanity_checks/test_performance_services.py]: If 0.9.0.X is the current baseline, maybe we don't need to run 0.8.2.X every time?

Comment (granders) [inline tests/kafkatest/sanity_checks/test_performance_services.py:40]: Maybe an additional comment here to justify why we would keep the 0.8.2.X test case around for the time being?

File: tests/kafkatest/services/performance/consumer_performance.py [hunk 6]
```diff
@@ -78,6 +78,11 @@ def __init__(self, context, num_nodes, kafka, topic, messages, new_consumer=Fals
         self.new_consumer = new_consumer
         self.settings = settings
 
+        if version < V_0_9_0_0 and self.new_consumer:
+            self.logger.warn("Version is %s where new_consumer is not available, but new_consumer is set to True. "\
+                             "Overriding new_consumer flag to False..." % str(version))
+            self.new_consumer = False
```
Comment (granders) [inline tests/kafkatest/services/performance/consumer_performance.py]: Looking back at this, it's a bit fishy - maybe better to enforce this with an assertion rather than warning and modifying the `self.new_consumer`

File: tests/kafkatest/services/performance/consumer_performance.py [hunk 7]
```diff
@@ -127,15 +135,31 @@ def start_cmd(self, node):
         cmd += " /opt/%s/bin/kafka-consumer-perf-test.sh" % kafka_dir(node)
         for key, value in self.args.items():
             cmd += " --%s %s" % (key, value)
-        cmd += " --consumer.config %s" % ConsumerPerformanceService.CONFIG_FILE
 
-        for key, value in self.settings.items():
-            cmd += " %s=%s" % (str(key), str(value))
+        if node.version > LATEST_0_8_2:
+            # This is only used for security settings
+            cmd += " --consumer.config %s" % ConsumerPerformanceService.CONFIG_FILE
```
Comment (granders) [inline tests/kafkatest/services/performance/consumer_performance.py:144]: On second look - `cmd += " --consumer.config %s" % ConsumerPerformanceService.CONFIG_FILE` was simply moved here

Comment (ijuma) [inline tests/kafkatest/services/performance/consumer_performance.py:144]: Yeah, that's why I only asked about the two lines copying the `settings`. So, I was going to add it back after this `if` block. Is that correct?

File: tests/kafkatest/services/performance/end_to_end_latency.py [hunk 8]
```diff
@@ -17,9 +17,11 @@
 from kafkatest.services.security.security_config import SecurityConfig
 
 from kafkatest.services.kafka.directory import kafka_dir
+from kafkatest.services.kafka.version import TRUNK, LATEST_0_8_2
 
 
 class EndToEndLatencyService(PerformanceService):
+    MESSAGE_BYTES = 21  # 0.8.X messages are fixed at 21 bytes, so we'll match that for other versions
```
Comment (granders) [inline tests/kafkatest/services/performance/end_to_end_latency.py:24]: It's probably worth double-checking whether this assertion is true

From what I remember, the 0.8.X version of EndToEndLatency sends a hard-coded message, and I computed it at 21 bytes, but it's possible that past Geoff made a mistake

File: tests/kafkatest/services/performance/end_to_end_latency.py [hunk 9]
```diff
@@ -36,28 +38,65 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, consumer_fetch
             'num_records': num_records,
             'consumer_fetch_max_wait': consumer_fetch_max_wait,
             'acks': acks,
-            'kafka_opts': self.security_config.kafka_opts
+            'kafka_opts': self.security_config.kafka_opts,
```
Comment (granders) [inline tests/kafkatest/services/performance/end_to_end_latency.py:46]: It may be worth validating that security protocol is plaintext if version is less than 0.9.0.0
Same comment for other services where it isn't already enforced

File: tests/kafkatest/services/performance/producer_performance.py (hunk 4 above)
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: After looking into the WARNINGs you mentioned collecting logs, it might make sense to set the jmx_log collect_default to True iff `jmx_object_names is not None`

Comment (ijuma) [inline tests/kafkatest/services/performance/producer_performance.py]: Thanks for looking into that, will do.

File: tests/kafkatest/services/performance/end_to_end_latency.py (hunk 8 above)
Comment (ijuma) [inline tests/kafkatest/services/performance/end_to_end_latency.py:24]: Past Geoff was correct.

**Comment by ijuma** [comment]
@granders, ready for another round of reviews. I addressed your feedback and added version support to `benchmark_test.py`. I don't have any other changes planned.

I tested the relevant tests via the AWS test driver and things seem fine. I also started a full system tests run via the branch builder:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder/407/

**Comment by ijuma** [comment]
System tests passed.

@ewencp, it would also be good for you to take a look as I assume you'll be the committer reviewing this. Some context: Geoff did most of the work on this branch before 0.9.0.0 went out. I needed this functionality, so Geoff rebased the branch and I tested it and added minor fixes and improvements.

File: tests/kafkatest/services/performance/producer_performance.py [hunk 10]
```diff
@@ -44,56 +74,103 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, record_size, t
...
             'bootstrap_servers': self.kafka.bootstrap_servers(self.security_config.security_protocol),
             'jmx_port': self.jmx_port,
             'client_id': self.client_id,
             'kafka_directory': kafka_dir(node)
             })
-        cmd = "JMX_PORT=%(jmx_port)d KAFKA_OPTS=%(kafka_opts)s /opt/%(kafka_directory)s/bin/kafka-run-class.sh org.apache.kafka.tools.ProducerPerformance " \
+
+        cmd = ""
+
+        if node.version <= LATEST_0_8_2:
```
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: @I think this should actually probably be:
`if node.version < TRUNK:`

Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py]: amended: make that strictly less than

**Comment by granders** [comment]
@ijuma Other than fixing the conditional in `producer_performance.py` this looks good to me. Let's see if @ewencp has further comments

File: tests/kafkatest/services/performance/producer_performance.py (hunk 10 above)
Comment (ijuma) [inline tests/kafkatest/services/performance/producer_performance.py]: Fixed it locally. Testing it before I push (as we discussed, I noticed this issue as well when running a performance test with a change to `ProducerPerformance`).

File: tests/kafkatest/services/performance/producer_performance.py [hunk 11]
```diff
@@ -44,56 +74,103 @@ def __init__(self, context, num_nodes, kafka, topic, num_records, record_size, t
...
-                'latency_max_ms': float(parts[3].split()[0]),
-                'latency_50th_ms': float(parts[4].split()[0]),
-                'latency_95th_ms': float(parts[5].split()[0]),
-                'latency_99th_ms': float(parts[6].split()[0]),
-                'latency_999th_ms': float(parts[7].split()[0]),
-            }
-        last = None
+        # start ProducerPerformance process
         producer_output = node.account.ssh_capture(cmd)
+        wait_until(lambda: self.alive(node), timeout_sec=20, err_msg="ProducerPerformance failed to start")
```
Comment (ewencp) [inline tests/kafkatest/services/performance/producer_performance.py:136]: @ijuma This seems like it might not work. In practice, all tests that use this should have it run long enough for this to pass (or correctly fail). But your sanity checks seem like a case where this could fail -- the process could startup and manage to exit before the first check for liveness runs.

File: tests/kafkatest/services/performance/templates/tools_log4j.properties [hunk 12]
```diff
@@ -19,7 +19,7 @@ log4j.rootLogger = {{ log_level|default("INFO") }}, FILE
 log4j.appender.FILE=org.apache.log4j.FileAppender
 log4j.appender.FILE.File={{ log_file }}
 log4j.appender.FILE.ImmediateFlush=true
-# Set the append to false, overwrite
-log4j.appender.FILE.Append=false
+# Set the append to true, append
+log4j.appender.FILE.Append=true
```
Comment (ewencp) [inline tests/kafkatest/services/performance/templates/tools_log4j.properties]: Why was this needed? Append is normal for long running services that we need to bounce since we want to capture all the output, but this seems unusual for tools. They are usually one-off. I don't think there's necessarily anything wrong with this, but I'd like to understand the reasoning since this template is used across 5 or 6 separate tools.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 11 above)
Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py:136]: @ewencp Isn't this true of any check of "aliveness" here since the process can be gone by the time you check?

In principle you'd need to detect
"is alive OR was alive and finished successfully"

One option is to simply keep as-is, but document that there a known (but probably unlikely) way to get a spurious failure here

File: tests/kafkatest/services/performance/templates/tools_log4j.properties (hunk 12 above)
Comment (granders) [inline tests/kafkatest/services/performance/templates/tools_log4j.properties]: I may have made this change in the original branch back in December, and I honestly don't remember if I had a good reason for it.

I agree it's probably safer to revert this since several tools rely on it.

File: tests/kafkatest/services/performance/producer_performance.py (hunk 11 above)
Comment (ewencp) [inline tests/kafkatest/services/performance/producer_performance.py:136]: @granders Yes, what you say is true, but in practice this type of test _normally_ works if you expect the process to be running for more than a few seconds. Normally I wouldn't have called this out, but since the sanity check tests that were added specifically try to be short-lived, I want to make sure we're not likely to encounter transient failures on them.

Comment (granders) [inline tests/kafkatest/services/performance/producer_performance.py:136]: Fair enough - my thought was that maybe it's ok to keep it simple (but document) for the time being since we aren't seeing these transient failures?

**Comment by ijuma** [comment]
@ewencp Running the sanity check test in a loop of 100 iterations here:

https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/49/consoleFull

So far, 20 iterations passed.

**Comment by ijuma** [comment]
OK, so the test passed 67 times and then it failed with what I think is an unrelated failure:

[INFO:2016-04-09 03:35:02,127]: SerialTestRunner: Summary: There aren't enough available nodes to satisfy the resource request. Total cluster size: 0, Requested: 1, Already allocated: 0, Available: 0. Make sure your cluster has enough nodes to run your test or service(s). Currently registered services: [<ZookeeperService: num_nodes: 1, nodes: []>]

**Comment by ewencp** [comment]
@ijuma That actually sounds like something went wrong with getting the list of machines available in the cluster (or they shutdown or something). This is what you normally get if you don't have any nodes started, or if you just don't have enough nodes for the test if the numbers are larger. 67 passes is enough to convince me :)

LGTM
